### ISSUE TITLE
Support for Custom opcodes in in-line assembly

### ISSUE THREAD

**Issue by SergioDemianLerner** [ethereum/solidity#3215]
For experimentation, we are creating new Ethereum opcodes. The Solidity documentation does not mention any way of embedding a custom opcode in the inline assembly code. 
Therefore new opcodes cannot be accessed from Solidity without modifying its source code. One temporal solution we're doing is calling the new opcodes through a "system trap" by overriding the CALL opcode. The result is something like:
```
    assembly {
            call(not(0x0),0,not(0x0), {new=opcode} , {new-opcode arg0}, {..arg1}, {..arg2})
            pop
        }
```
Even this works, it's ugly. It would be very helpful that the assembly adds a meta-opcode:
```
inNoutM(opcode,args....)
```

Therefore an opcode with code 0xaf with 2 inputs arguments and 1 output argument can be called like this:
```
r :=in2out1(0xaf, 0x01, 0x02)
```
An opcode 0xae with 1 input an zero outputs like this
```
in1out0(0xae,0x01)
```

I offer to create the code and do a pull request in order to enable this functionality. 
Also maybe there is a better cleaner alternative I haven't thought about. Maybe one could just define the new opcode with a syntax like: 
```
  define(0xaf,2,1,ERASE)
  ...
  r := erase(0x01,0x02)
```

**Comment by axic** [comment]
I like the idea of defining new opcodes, but unfortunately this goes straight in contrast to the planned evolution of the assembly language (see #2129), where no instructions (opcodes) are present only functions and those are defined by the backend.

Do you have any suggestions how to fit this into such a system?

Also it is questionable how many and frequently new opcodes should be added to the EVM and if that warrants a system to include new opcodes on the fly as opposed to adding them as 3 new lines in to the C++ source code. Is that such a big overhead considering they must also be implemented in clients?

**Comment by chriseth** [comment]
This is something that would be really useful. The problem is that it is not really portable, but something like the following could be a good compromise:

```
function newOpcode(a, b, c) -> d = EVM1 { 0xab }
```

So basically you only define how it will be implemented for EVM 1.0 and fail for the other backends

**Comment by axic** [comment]
@chriseth where this becomes quite complicated is supporting multi-byte opcodes or opcodes with immediate arguments (e.g. the SIMD proposal).

**Comment by chriseth** [comment]
@axic yes, so perhaps it is better to keep the requirement for compiler modification but make it easy by providing a set of built-in functions, so what we were planning anyway ;-)

**Comment by SergioDemianLerner** [comment]
Any option is good for me.  I think It's not so important if the solution we pick today will be forever supported in the language. If Solidity migrates to another way of defining custom opcodes, then users can very easily change that part of the code. Any use of custom opcodes will be generally encapsulated in higher level solidity functions. Upgrading contracts will be easy.

**Comment by federicobond** [comment]
I am more inclined to add a function/meta-instruction but making it really clear that it's not an "official" part of the language API, just a workaround for some use cases. If we can infer the in-out argument count we could perhaps do something like this:
```
r := unsafeExecuteOpcode(0x67, a, b)
```

**Comment by SergioDemianLerner** [comment]
Would you accept a pull request with an implementation of this in one of the possible ways? I want to know if it's worth working on it.

**Comment by axic** [comment]
I am still not sure it is worth supporting a soon to be deprecated construct given it takes so long to make breaking releases.

Have you tried adding new opcodes to C++? If you have found any issues with that let us make it easier to add new ones to C++ as a solution. After all, whoever needs new opcodes will need to adjust client implementations too with similar level of complexity.

**Comment by axic** [comment]
This has been discussed and eventually implemented in #10869.

**Comment by 1m1-github** [comment]
hi ~ is there a guide on how to add an OPCODE to Solidity C++? ty
